usr.bin/script/script.1
.\"	$NetBSD: script.1,v 1.10 2003/08/07 11:15:48 agc Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)script.1	8.1 (Berkeley) 6/6/93
.\"
.Dd June 6, 1993
.Dt SCRIPT 1
.Os
.Sh NAME
.Nm script
.Nd make typescript of terminal session
.Sh SYNOPSIS
.Nm
.Op Fl adpr
.Op Fl c Ar command
.Op Ar file
.Sh DESCRIPTION
.Nm
makes a typescript of everything printed on your terminal.
It is useful for students who need a hardcopy record of an interactive
session as proof of an assignment, as the typescript file
can be printed out later with
.Xr lpr 1 .
.Pp
If the argument
.Ar file
is given,
.Nm
saves all dialogue in
.Ar file .
If no file name is given, the typescript is saved in the file
.Pa typescript  .
.Pp
Option:
.Bl -tag -width Ds
.It Fl a
Append the output to
.Ar file
or
.Pa typescript ,
retaining the prior contents.
.It Fl d
Don't sleep between records when playing back a timestamped session.
.It Fl p
Play back a recorded session in real time.
.It Fl r
Record a session with input, output, and timestamping.
.It Fl c Ar command
Run
.Ar command
instead of a shell.  This works only if the shell found via
.Dv \&$SHELL
(see below) understands the
.Fl c
convention for passing a command to execute, as the usual shells do.
.El
.Pp
The script ends when the forked shell exits (the EOF character,
typically
.Em control-D ,
to exit
the Bourne shell
.Pf ( Xr sh 1 ) ;
.Em exit ,
.Em logout
or (if
.Em ignoreeof
is not set) the EOF character for the
C-shell,
.Xr csh 1 ) .
.Pp
Certain interactive commands, such as
.Xr vi 1 ,
create garbage in the typescript file.
.Nm
works best with commands that do not manipulate the
screen; the results are meant to emulate a hardcopy
terminal.
.Sh ENVIRONMENT
The following environment variable is used by
.Nm :
.Bl -tag -width SHELL
.It Ev SHELL
If the variable
.Ev SHELL
exists, the shell forked by
.Nm
will be that shell. If
.Ev SHELL
is not set, the Bourne shell
is assumed.  (Most shells set this variable automatically).
.El
.Sh SEE ALSO
.Xr csh 1
(for the
.Em history
mechanism).
.Sh HISTORY
The
.Nm
command appeared in
.Bx 3.0 .
.Sh BUGS
.Nm
places
.Sy everything
in the log file, including linefeeds and backspaces.
This is not what the naive user expects.
.Pp
When not using
.Fl c ,
the code assumes
.Dv \&$SHELL
accepts
.Fl i
to indicate
.Sq interactive .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1142.2.10 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more NuBus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	puc(4): Add support for Perle Systems 4 and 8 ports modem boards.
		[manu 20071209]
	arm: Add support for NPE Ethernet on IXP425. [scw 20061210]
	arm: Add support for IXP425 watchdog timer. [scw 20061210]
	libutil: Add flags_to_string(3) and string_to_flags(3),
		moved from ls(1). [christos 20061214]
	arm: Add support for i2c, i2s, pcmcia, and ohci of PXA2x0.
		Ported from OpenBSD as part of zaurus port by NONAKA Kimihiro.
		[peter 20061217]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20061217]
	kernel: Merge yamt-splraseipl branch, which provides splraiseipl(9)
		and makeiplcookie(9). [yamt 20061221]
	sgimips: Add support for Set Engineering's GIO Fast Ethernet board.
		[rumble 20061223]
	sgimips: Add support for the SGI Light/Entry/Starter LG1/LG2
		framebuffers found in Indigo systems. [rumble 20061226]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	kernel, srtconfig(1): First import of the source-address-based routing
		pseudo-device.  [mouse 20061229]
	sgimips: Add i8254 timecounter(9) support for use on IP12.
		[rumble 20061229]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	njata(4): Add support for Workbit CF32A CF adapter,
		aka BUFFALO RCF-CBA2.  [ichiro 20061231]
	amd64, i386: Report and work around some of the known errata for
		Athlon 64 and Opteron processors. [ad 20070101]
	i386: Add fatboot, sub-512 byte bootloader that can be written into
		the first sector of a FAT16 filesystem and will load
		the secondary bootloader from it. [dsl 20070101]
	kernel: Make mount(2) and unmount(2) use kauth(9) for security
		policy.  One incompatible change is that an unprivileged
		mount without the MNT_NOSUID and MNT_NODEV flags
		will now fail, whereas previously those flags were
		automatically added.  [elad 20070102]
	udf: Implement read-only access to UDF 2.50 and UDF 2.60 discs
		that use metadata partition mappings. [reinoud 20070104]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	config(1): Introduce a new -L option, generates a lint configuration.
		[cube 20070108]
	umass(4): Add a quirk for the Sony Cybershot USB camera.
		From Mihai Chelaru. [christos 20070109]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	zaurus: Add MI todr(9) and timecounter(9) support. [ober 20070112]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	kernel: Add syscalls for POSIX Real-time library. [ad 20070116]
	adb(4): Add new and simplified MI ADB drivers. [macallan 20070117]
	macppc: Add a driver for the SGS 7433 mixer found in some G3 Macs.
		[macallan 20070117]
	viaenv(4): Add support for the VIA VT8231 hardware monitor and
		remove viapm(4) code. From OpenBSD. [xtraeme 20070120]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	msk(4): Pull support for some devices and many bug fixes from OpenBSD.
		[msaitoh 20070130]
	uplcom(4): Add support for Willcom WS002IN PHS device
		(Prolific Technology PL2303X). [msaitoh 20070131]
	zaurus: Updates to allow Zaurus screen to rotate 90 degrees to a usable
		state with the keyboard. From OpenBSD. [ober,peter 20070202]
	security(8): Add support for per-user /tmp. [elad 20070204]
	ugensa(4): Add support Novatel Merlin, used in VZ V620.
		From John Nielsen on freebsd-mobile. [gdt 20070205]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	find(1): Add -delete option from FreeBSD. [elad 20070206]
	rum(4): Add support for MELCO WLI-U2-SG54HP and PLANEX GW-US54Mini2.
		[elad 20070206]
	kernel: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	ubsa(4): Add support for CDMA modems sold by Eurotel (now O2).
		[salo 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	shark: Add generic softintr(9) support. [matt 20070218]
	kernel: Add code to support per-system call statistics. [dsl 20070218]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	sgimips: Support the sq(4) interface on the Challenge S's IOPLUS

usr.bin/script/script.c
/*	$NetBSD: script.c,v 1.16 2008/07/21 14:19:25 lukem Exp $	*/

/*
 * Copyright (c) 1980, 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1992, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)script.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: script.c,v 1.16 2008/07/21 14:19:25 lukem Exp $");
#endif /* not lint */

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/param.h>
#include <sys/uio.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <tzfile.h>
#include <unistd.h>
#include <util.h>

#define	DEF_BUF	65536

struct stamp {
	uint64_t scr_len;	/* amount of data */
	uint64_t scr_sec;	/* time it arrived in seconds... */
	uint32_t scr_usec;	/* ...and microseconds */
	uint32_t scr_direction;	/* 'i', 'o', etc (also indicates endianness) */
};

FILE	*fscript;
int	master, slave;
int	child, subchild;
int	outcc;
int	usesleep, rawout;
char	*fname;
static char *cmd = 0;

struct	termios tt;

void	done(void);
void	dooutput(void);
void	docmd(void);
void	fail(void);
void	finish(int);
int	main(int, char **);
void	scriptflush(int);
void	record(FILE *, char *, size_t, int);
void	consume(FILE *, off_t, char *, int);
void	playback(FILE *);

int
main(int argc, char *argv[])
{
	int cc;
	struct termios rtt;
	struct winsize win;
	int aflg, pflg, ch;
	char ibuf[BUFSIZ];

	aflg = 0;
	pflg = 0;
	usesleep = 1;
	rawout = 0;
	while ((ch = getopt(argc, argv, "ac:dpr")) != -1)
		switch(ch) {
		case 'a':
			aflg = 1;
			break;
		case 'c':
			cmd = optarg;
			break;
		case 'd':
			usesleep = 0;
			break;
		case 'p':
			pflg = 1;
			break;
		case 'r':
			rawout = 1;
			break;
		case '?':
		default:
			(void)fprintf(stderr, "usage: %s [-adpr] [file]\n",
			    getprogname());
			exit(1);
		}
	argc -= optind;
	argv += optind;

	if (argc > 0)
		fname = argv[0];
	else
		fname = "typescript";

	if ((fscript = fopen(fname, pflg ? "r" : aflg ? "a" : "w")) == NULL)
		err(1, "fopen %s", fname);

	if (pflg)
		playback(fscript);

	(void)tcgetattr(STDIN_FILENO, &tt);
	(void)ioctl(STDIN_FILENO, TIOCGWINSZ, &win);
	if (openpty(&master, &slave, NULL, &tt, &win) == -1)
		err(1, "openpty");

	(void)printf("Script started, output file is %s\n", fname);
	rtt = tt;
	cfmakeraw(&rtt);
	rtt.c_lflag &= ~ECHO;
	(void)tcsetattr(STDIN_FILENO, TCSAFLUSH, &rtt);

	(void)signal(SIGCHLD, finish);
	child = fork();
	if (child < 0) {
		warn("fork");
		fail();
	}
	if (child == 0) {
		subchild = child = fork();
		if (child < 0) {
			warn("fork");
			fail();
		}
		if (child)
			dooutput();
		else
			docmd();
	}

	if (!rawout)
		(void)fclose(fscript);
	while ((cc = read(STDIN_FILENO, ibuf, BUFSIZ)) > 0) {
		if (rawout)
			record(fscript, ibuf, cc, 'i');
		(void)write(master, ibuf, cc);
	}
	done();
	/* NOTREACHED */
	return (0);
}

void
finish(int signo)
{
	int die, pid, status;

	die = 0;
	while ((pid = wait3(&status, WNOHANG, 0)) > 0)
		if (pid == child)
			die = 1;

	if (die)
		done();
}

void
dooutput()
{
	struct itimerval value;
	int cc;
	time_t tvec;
	char obuf[BUFSIZ];

	(void)close(STDIN_FILENO);
	tvec = time(NULL);
	if (rawout)
		record(fscript, NULL, 0, 's');
	else
		(void)fprintf(fscript, "Script started on %s", ctime(&tvec));

	(void)signal(SIGALRM, scriptflush);
	value.it_interval.tv_sec = SECSPERMIN / 2;
	value.it_interval.tv_usec = 0;
	value.it_value = value.it_interval;
	(void)setitimer(ITIMER_REAL, &value, NULL);
	for (;;) {
		cc = read(master, obuf, sizeof (obuf));
		if (cc <= 0)
			break;
		(void)write(1, obuf, cc);
		if (rawout)
			record(fscript, obuf, cc, 'o');
		else
			(void)fwrite(obuf, 1, cc, fscript);
		outcc += cc;
	}
	done();
}

void
scriptflush(int signo)
{
	if (outcc) {
		(void)fflush(fscript);
		outcc = 0;
	}
}

void
docmd()
{
 char *shell;

 shell = getenv("SHELL");
 if (! shell) shell = _PATH_BSHELL;
 close(master);
 fclose(fscript);
 login_tty(slave);
 if (cmd)
  { execl(shell,shell,"-c",cmd,(char *)0);
  }
 else
  { execl(shell,shell,"-i",(char *)0);
  }
 warn("execl %s",shell);
 fail();
}

void
fail()
{

	(void)kill(0, SIGTERM);
	done();
}

void
done()
{
	time_t tvec;

	if (subchild) {
		tvec = time(NULL);
		if (rawout)
			record(fscript, NULL, 0, 'e');
		else
			(void)fprintf(fscript,"\nScript done on %s",
			    ctime(&tvec));
		(void)fclose(fscript);
		(void)close(master);
	} else {
		(void)tcsetattr(STDIN_FILENO, TCSAFLUSH, &tt);
		(void)printf("Script done, output file is %s\n", fname);
	}
	exit(0);
}

void
record(FILE *fscript, char *buf, size_t cc, int direction)
{
	struct iovec iov[2];
	struct stamp stamp;
	struct timeval tv;

	(void)gettimeofday(&tv, NULL);
	stamp.scr_len = cc;
	stamp.scr_sec = tv.tv_sec;
	stamp.scr_usec = tv.tv_usec;
	stamp.scr_direction = direction;
	iov[0].iov_len = sizeof(stamp);
	iov[0].iov_base = &stamp;
	iov[1].iov_len = cc;
	iov[1].iov_base = buf;
	if (writev(fileno(fscript), &iov[0], 2) == -1)
		err(1, "writev");
}

void
consume(FILE *fscript, off_t len, char *buf, int reg)
{
	size_t l;

	if (reg) {
		if (fseeko(fscript, len, SEEK_CUR) == -1)
			err(1, NULL);
	}
	else {
		while (len > 0) {
			l = MIN(DEF_BUF, len);
			if (fread(buf, sizeof(char), l, fscript) != l)
				err(1, "cannot read buffer");
			len -= l;
		}
	}
}

#define swapstamp(stamp) do { \
	if (stamp.scr_direction > 0xff) { \
		stamp.scr_len = bswap64(stamp.scr_len); \
		stamp.scr_sec = bswap64(stamp.scr_sec); \
		stamp.scr_usec = bswap32(stamp.scr_usec); \
		stamp.scr_direction = bswap32(stamp.scr_direction); \
	} \
} while (0/*CONSTCOND*/)

void
playback(FILE *fscript)
{
	struct timespec tsi, tso;
	struct stamp stamp;
	struct stat pst;
	char buf[DEF_BUF];
	off_t nread, save_len;
	size_t l;
	time_t clock;
	int reg;

	if (fstat(fileno(fscript), &pst) == -1)
		err(1, "fstat failed");	

	reg = S_ISREG(pst.st_mode);

	for (nread = 0; !reg || nread < pst.st_size; nread += save_len) {
		if (fread(&stamp, sizeof(stamp), 1, fscript) != 1) {
			if (reg)
				err(1, "reading playback header");
			else
				break;
		}
		swapstamp(stamp);
		save_len = sizeof(stamp);

		if (reg && stamp.scr_len >
		    (uint64_t)(pst.st_size - save_len) - nread)
			err(1, "invalid stamp");

		save_len += stamp.scr_len;
		clock = stamp.scr_sec;
		tso.tv_sec = stamp.scr_sec;
		tso.tv_nsec = stamp.scr_usec * 1000;

		switch (stamp.scr_direction) {
		case 's':
			(void)printf("Script started on %s", ctime(&clock));
			tsi = tso;
			(void)consume(fscript, stamp.scr_len, buf, reg);
			break;
		case 'e':
			(void)printf("\nScript done on %s", ctime(&clock));
			(void)consume(fscript, stamp.scr_len, buf, reg);
			break;
		case 'i':
			/* throw input away */
			(void)consume(fscript, stamp.scr_len, buf, reg);
			break;
		case 'o':
			tsi.tv_sec = tso.tv_sec - tsi.tv_sec;
			tsi.tv_nsec = tso.tv_nsec - tsi.tv_nsec;
			if (tsi.tv_nsec < 0) {
				tsi.tv_sec -= 1;
				tsi.tv_nsec += 1000000000;
			}
			if (usesleep)
				(void)nanosleep(&tsi, NULL);
			tsi = tso;
			while (stamp.scr_len > 0) {
				l = MIN(DEF_BUF, stamp.scr_len);
				if (fread(buf, sizeof(char), l, fscript) != l)
					err(1, "cannot read buffer");

				(void)write(STDOUT_FILENO, buf, l);
				stamp.scr_len -= l;
			}
			break;
		default:
			err(1, "invalid direction");
		}
	}
	(void)fclose(fscript);
	exit(0);
}
